### ISSUE TITLE
L17_Image_Classification_MNIST.ipynb

### ISSUE THREAD

**Issue by uzinny04** [uzinny04/AI_Appl#1]
Image Classification (MNIST)
Imports

import torch
import torchvision.datasets as dsets
import torchvision.transforms as transforms
import torch.nn.init
from torch.utils.data import DataLoader

import random
import matplotlib.pyplot as plt

device = 'cuda' if torch.cuda.is_available() else 'cpu'
print(device)

# for reproducibility
torch.manual_seed(777)
if device == 'cuda':
    torch.cuda.manual_seed_all(777)
     
cuda
1 MNIST dataset

# MNIST dataset
mnist_train = dsets.MNIST(root='MNIST_data/',
                          train=True,
                          transform=transforms.ToTensor(),
                          download=True)

mnist_test = dsets.MNIST(root='MNIST_data/',
                         train=False,
                         transform=transforms.ToTensor(),
                         download=True)
     
Downloading http://yann.lecun.com/exdb/mnist/train-images-idx3-ubyte.gz
Failed to download (trying next):
HTTP Error 403: Forbidden

Downloading https://ossci-datasets.s3.amazonaws.com/mnist/train-images-idx3-ubyte.gz
Downloading https://ossci-datasets.s3.amazonaws.com/mnist/train-images-idx3-ubyte.gz to MNIST_data/MNIST/raw/train-images-idx3-ubyte.gz
100%|██████████| 9912422/9912422 [00:00<00:00, 17901167.12it/s]
Extracting MNIST_data/MNIST/raw/train-images-idx3-ubyte.gz to MNIST_data/MNIST/raw

Downloading http://yann.lecun.com/exdb/mnist/train-labels-idx1-ubyte.gz
Failed to download (trying next):
HTTP Error 403: Forbidden

Downloading https://ossci-datasets.s3.amazonaws.com/mnist/train-labels-idx1-ubyte.gz
Downloading https://ossci-datasets.s3.amazonaws.com/mnist/train-labels-idx1-ubyte.gz to MNIST_data/MNIST/raw/train-labels-idx1-ubyte.gz
100%|██████████| 28881/28881 [00:00<00:00, 515824.43it/s]
Extracting MNIST_data/MNIST/raw/train-labels-idx1-ubyte.gz to MNIST_data/MNIST/raw

Downloading http://yann.lecun.com/exdb/mnist/t10k-images-idx3-ubyte.gz
Failed to download (trying next):
HTTP Error 403: Forbidden

Downloading https://ossci-datasets.s3.amazonaws.com/mnist/t10k-images-idx3-ubyte.gz
Downloading https://ossci-datasets.s3.amazonaws.com/mnist/t10k-images-idx3-ubyte.gz to MNIST_data/MNIST/raw/t10k-images-idx3-ubyte.gz
100%|██████████| 1648877/1648877 [00:00<00:00, 1925518.73it/s]
Extracting MNIST_data/MNIST/raw/t10k-images-idx3-ubyte.gz to MNIST_data/MNIST/raw

Downloading http://yann.lecun.com/exdb/mnist/t10k-labels-idx1-ubyte.gz
Failed to download (trying next):
HTTP Error 403: Forbidden

Downloading https://ossci-datasets.s3.amazonaws.com/mnist/t10k-labels-idx1-ubyte.gz
Downloading https://ossci-datasets.s3.amazonaws.com/mnist/t10k-labels-idx1-ubyte.gz to MNIST_data/MNIST/raw/t10k-labels-idx1-ubyte.gz
100%|██████████| 4542/4542 [00:00<00:00, 3221804.29it/s]
Extracting MNIST_data/MNIST/raw/t10k-labels-idx1-ubyte.gz to MNIST_data/MNIST/raw

# dataset loader
batch_size = 100
data_loader = DataLoader(dataset=mnist_train,
                                          batch_size=batch_size,
                                          shuffle=True,
                                          drop_last=True)
     
2 CNN model
Input (입력의 형태)
Input type: torch.Tensor
Input shape: (N x C x H x W)
N: Batch size, C: # of channels, H: height, W: width
(?, 1, 28, 28)
여러장의, 흑백, 28x28 size의 이미지라고 가정하자
Layer 설계
Layer 1
Conv2d >> C: 32, Kernel size (필터 크기): 3, Stride: 1, Padding: 1
ReLU
MaxPool >> Kernel size: 2, Stride: 2
입-출력 (?, 1, 28, 28) >> (?, 32, 14, 14)
Layer 2
Conv2d >> C: 64, Kernel size (필터 크기): 3, Stride: 1, Padding: 1
ReLU
MaxPool >> Kernel size: 2, Stride: 2
입-출력 (?, 32, 14, 14) >> (?, 64, 7, 7)
Layer 3
Linear >> input: 7x7x64 output: 10
Softmax

class CNN(torch.nn.Module):

    def __init__(self):
        super(CNN, self).__init__()
        # L1 Input shape=(?, 1, 28, 28)
        #    Conv     -> (?, 32, 28, 28)
        #    Pool     -> (?, 32, 14, 14)
        self.layer1 = torch.nn.Sequential(
            torch.nn.Conv2d(1, 32, kernel_size=3, stride=1, padding=1),
            torch.nn.ReLU(),
            torch.nn.MaxPool2d(kernel_size=2, stride=2))
        # L2 Input shape=(?, 32, 14, 14)
        #    Conv      ->(?, 64, 14, 14)
        #    Pool      ->(?, 64, 7, 7)
        self.layer2 = torch.nn.Sequential(
            torch.nn.Conv2d(32, 64, kernel_size=3, stride=1, padding=1),
            torch.nn.ReLU(),
            torch.nn.MaxPool2d(kernel_size=2, stride=2))
        # L3 Linear 64x7x7 inputs -> 10 outputs
        self.fc = torch.nn.Linear(64*7*7, 10, bias=True)
        torch.nn.init.xavier_uniform_(self.fc.weight)

    def forward(self, x):
        out = self.layer1(x)
        out = self.layer2(out)
        out = out.view(out.size(0), -1)   # Flatten them for FC
        out = self.fc(out)
        return out
     
3 학습하기import torch
import torch.cuda as cuda
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torch.utils.data import DataLoader, Dataset
from torchvision.transforms import RandomCrop, Resize, Compose, ToTensor, Normalize
from torchvision.datasets import ImageFolder
device = 'cuda' if cuda.is_available() else 'cpu'
print(device)

torch.manual_seed(1)
if device=='cuda':
    torch.cuda.manual_seed_all(1)
from google.colab import drive
drive.mount('/content/drive')##  4/1AX4XfWhBuMN--7VPIqiRFsSTKJdLT6OjelvFJFyw2vWmjpSmPeUFoJLhmZA
class mydataset(Dataset):
    def __init__(self, path, transfer):
        data = ImageFolder(root=path, transform=transfer)
        self.imgs = []
        self.labels = []

        for i, (img, label) in enumerate(data):
            self.imgs.append(img)
            self.labels.append([label])

        self.length = len(self.labels)
    
    def __len__(self):
        return self.length
    
    def __getitem__(self, item):
        data = self.imgs[item]
        target = torch.Tensor(self.labels[item]).to(torch.long).squeeze()
        return data, target

transferF = Compose([Resize([64, 128]), RandomCrop([64, 128]), ToTensor()])
transferFte = Compose([Resize([64, 128]), ToTensor()])

train_dataset = mydataset('/content/drive/MyDrive/AI_Appl/train_data', transferF)

train_loader = torch.utils.data.DataLoader(dataset=train_dataset,
                                           batch_size=15,
                                           shuffle=True,)

test_dataset = mydataset('/content/drive/MyDrive/AI_Appl/test_data', transferFte)
test_loader = torch.utils.data.DataLoader(dataset=test_dataset,
                                           batch_size=15,
                                           shuffle=True)
class Net(nn.Module):

    def __init__(self):
        super(Net, self).__init__()
        self.imagesize = (64,128,3)

        self.conv1 = nn.Conv2d(3, 32, kernel_size=3, padding=1)     #out 10,64,128
        self.conv2 = nn.Conv2d(32, 64, kernel_size=3, padding=1)    #out 20,64,128
        self.mp1 = nn.MaxPool2d(2)                                   #out 20,16,32

        self.conv3 = nn.Conv2d(64, 128, kernel_size=3, padding=1)  # out 20,8,16
        self.conv4 = nn.Conv2d(128, 256, kernel_size=3, padding=1)  # out 20,4,8
        self.mp2 = nn.MaxPool2d(2)  # out 20,16,32

        self.fc1 = nn.Linear(256 * 4 * 8, 512)
        self.fc2 = nn.Linear(512, 2)

    def forward(self, x):
        in_size = x.size(0)

        x = F.relu(self.mp1(self.conv1(x)))
        x = F.relu(self.mp1(self.conv2(x)))

        x = F.relu(self.mp1(self.conv3(x)))
        x = F.relu(self.mp2(self.conv4(x)))

        x = x.view(in_size, -1)  # flatten the tensor
        x = self.fc1(x)
        x = self.fc2(x)             # batch, 4

        return x

model = Net()
model = model.to(device)
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model.parameters(), lr=0.00003)
Best = -1

for epoch in range(1, 11):
  
    train_loss = 0
    for batch_idx, (data, target) in enumerate(train_loader):
        data = data.to(device)
        target = target.to(device)
        output = model(data)
        loss = criterion(output, target)
        train_loss += loss

        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
    with torch.no_grad():
      test_loss = 0
      correct = 0
      
      for data, target in test_loader:
          data = data.to(device)
          target = target.to(device)
          output = model(data)
          test_loss += criterion(output, target).item()
          pred = output.data.max(1, keepdim=True)[1]
          correct += pred.eq(target.data.view_as(pred)).cpu().sum()

      acc = 100. * correct / len(test_loader.dataset)
      if acc > Best:
          Best = acc
          
      print('\nTrain set {} Epoch: loss: {:.8f}'.format(epoch, loss))
      print('\nTest set:  Accuracy: {}/{} ({:.0f}%), Best Accuracy {:.0f}%\n'.format(
           correct, len(test_loader.dataset), acc, Best))
      print('=='*100)
    
     
cuda
Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
Learning started. It takes sometime.
[Epoch:    1] cost = 0.158599436
[Epoch:    2] cost = 0.00020430611
[Epoch:    3] cost = 2.68744279e-05
[Epoch:    4] cost = 5.39911434e-06
[Epoch:    5] cost = 1.20812297e-06
Learning Finished!
Accuracy: 1.0
